### PR TITLE
NGWMN-2099 -- Added json output to /data end point.

### PR DESCRIPTION
**Pull request by orosz-usgs**
Updated groundwaterDailyValue.xml and dao to accept bestTS
parameter
  New test: DataControllerJsonIT
  Made monitoringLocationID query parameter required.

Before making a pull request
----------------------------

- [ ] Make sure all tests run
- [ ] Update the changelog appropriately
- [ ] Make sure the Swagger docs are accurate

Title
-----------
Brief description of changes. Reference the JIRA ticket if appropriate

Description
-----------
If no JIRA ticket is referenced, describe the changes made. Note anything that you want the reviewers to know while
reviewing your pull request

After making a pull request
---------------------------
- [ ] If appropriate, put the link to the PR in the JIRA ticket
- [ ] Assign someone to review unless the change is trivial

### CODE REVIEW

File: src/main/java/gov/usgs/wma/waterdata/data/DataController.java [hunk 1]
```diff
@@ -59,14 +59,22 @@ public void getTimeSeries(
 					@ExampleObject(name = "waterML", value = "WaterML", description = "Water ML") }) @RequestParam(value = "f", required = false, defaultValue = "waterml") String mimeType,
 			HttpServletResponse response) throws HttpMediaTypeNotAcceptableException, IOException {
 
-		// not using the ContentType returned yet, since we only have waterML
-		determineContentType(mimeType, List.of(ContentType.waterml));
+		ContentType contentType = determineContentType(mimeType, List.of(ContentType.json, ContentType.waterml));
```
Comment (duselmann) [inline src/main/java/gov/usgs/wma/waterdata/data/DataController.java:62]: It might be better to make a CONTENT_TYPES_SUPPORTED const.

File: src/main/java/gov/usgs/wma/waterdata/data/DataController.java [hunk 2]
```diff
@@ -59,14 +59,22 @@ public void getTimeSeries(
...
 		String rtn = null;
 		String bestTS = best == null ? CollectionParams.PARAM_MATCH_ANY : best.toString().toLowerCase();
 
+		if (contentType.isJson() && !bestTS.equals("true")) {
+			throw new HttpMediaTypeNotAcceptableException("Json content is only available with parameter best=true");
```
Comment (duselmann) [inline src/main/java/gov/usgs/wma/waterdata/data/DataController.java:67]: Why is this condition part of the contract?

File: src/main/java/gov/usgs/wma/waterdata/data/DataController.java [hunk 3]
```diff
@@ -59,14 +59,22 @@ public void getTimeSeries(
...
 		if (Domain.includesGroundWaterLevels(domains) && dataType.isStatisticalTimeSeries()) {
-			rtn = timeSeriesDao.getTimeSeriesWaterML(monLocIdentifier, bestTS);
-			response.setContentType(MediaType.APPLICATION_XML_VALUE);
+			if(contentType.isJson()) {
+				rtn = timeSeriesDao.getTimeSeries(monLocIdentifier, bestTS);
+				response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+			} else {
+				rtn = timeSeriesDao.getTimeSeriesWaterML(monLocIdentifier, bestTS);
+				response.setContentType(MediaType.APPLICATION_XML_VALUE);
+			}
```
Comment (duselmann) [inline src/main/java/gov/usgs/wma/waterdata/data/DataController.java:77]: If we ever had more content types then this should be polymorphic. Say a enum that contained the rules or linked to a class that contained the rules. Long if-else chains are hard to follow.

File: src/main/resources/mybatis/groundwaterDailyValue.xml [hunk 4]
```diff
@@ -22,15 +22,30 @@
...
-                     and monitoring_location_identifier = #{featureId,jdbcType=VARCHAR}
+                       from nwis.groundwater_daily_value gdv
+                   left join (
+                      select monitoring_location_identifier, time_series_unique_id best_ts_id
+                      from (select *, RANK() OVER (ORDER BY ranking DESC) AS rk
+                      from nwis.best_time_series
+                      where monitoring_location_identifier = #{featureId,jdbcType=VARCHAR}
+                            and ranking > 0) as t
+                      where rk = 1) bts
+                    on gdv.monitoring_location_identifier = bts.monitoring_location_identifier
```
Comment (duselmann) [inline src/main/resources/mybatis/groundwaterDailyValue.xml:33]: nice join

**Review by duselmann — APPROVED** [review]
I have a couple future proofing suggestions but it looks good overall.

File: src/main/java/gov/usgs/wma/waterdata/data/DataController.java (hunk 2 above)
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/data/DataController.java:67]: Currently, the json (omsf) is not row based, it only has room for one observed_property (pcode) in the properties object.
Even if the extra pcode was stored overloaded in the properties, there would be no way to map them to the one result array. Hence, limiting to one time series (best=true),

Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/data/DataController.java:67]: good question.  This is certainly worth of a comment. I'll add some notes.

File: src/main/java/gov/usgs/wma/waterdata/data/DataController.java (hunk 3 above)
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/data/DataController.java:77]: Totally agree that long if chains are hard to follow, essentially when the code blocks are dozens/hundreds of lines long.
In this case, it is hard to say how many more content type there will be. There will be csv at some point.
This should be part of the refactor to using the accept headers for the content type.
